granulate_utils/linux/cgroups/cgroup.py

```diff
@@ -78,7 +95,99 @@ def find_v2_hierarchy(resolve_host_root_links: bool = True) -> Optional[str]:
     return path
 
 
-def get_cgroup_mount(controller: str, resolve_host_root_links: bool = True) -> Optional[str]:
+CGROUP_PROCS_FILE = "cgroup.procs"
+
+
+class CgroupCore:
+    path: Path
+
+    def __init__(self, path: Path):
+        self.path = path
+
+    def _get_parent_cgroup(self, cgroup_name: str) -> Optional[Path]:
+        cgroup_index = self.path.parts.index(cgroup_name)
+        if cgroup_index != -1:
+            return Path(*self.path.parts[: cgroup_index + 1])
+        return None
+
+    def _create_subcgroup(self, cgroup_name: str) -> Optional[Path]:
+        if self.has_parent_cgroup(cgroup_name):
+            new_cgroup_path = self._get_parent_cgroup(cgroup_name)
+        else:
+            new_cgroup_path = self.path / cgroup_name
+            new_cgroup_path.mkdir(exist_ok=True)
+
+        return new_cgroup_path
+
+    def has_parent_cgroup(self, cgroup_name: str) -> bool:
+        """
+        Check if cgroup has the given cgroup_name in its hierarchy
+        :param cgroup_name: the cgroup name
+        :return: bool indicating if cgroup is in hierarchy
+        """
+        controller_path_parts = self.path.parts
+        return cgroup_name in controller_path_parts
+
+    def get_pids_in_cgroup(self) -> set[int]:
+        return {int(proc) for proc in self.read_from_interface_file(CGROUP_PROCS_FILE).split()}
+
+    def assign_process_to_cgroup(self, pid: int = 0) -> None:
+        if not self.path.exists():
+            raise FileNotFoundError("Cgroup doesn't exist")
+
+        self.write_to_interface_file(CGROUP_PROCS_FILE, str(pid))
+
+    def read_from_interface_file(self, interface_name: str) -> str:
+        interface_path = self.path / interface_name
+        return interface_path.read_text()
+
+    def write_to_interface_file(self, interface_name: str, data: str) -> None:
+        interface_path = self.path / interface_name
+        interface_path.write_text(data)
+
+    @abstractmethod
+    def create_subcgroup(self, controller: str, cgroup_name: str) -> CgroupCore:
+        """
+        Create a new CGroup lower in the hierarchy.
+        If a cgroup with the given name is already in hierarchy, return its path.
+        """
+        pass
+
+
+class CgroupCoreV1(CgroupCore):
+    def create_subcgroup(self, controller: str, cgroup_name: str) -> CgroupCore:
+        subcgroup_path = self._create_subcgroup(cgroup_name)
+        assert subcgroup_path is not None
+        return CgroupCoreV1(subcgroup_path)
+
+
+CGROUP_SUPPORTED_CONTROLLERS = "cgroup.controllers"
+CGROUP_ENABLED_CONTROLLERS = "cgroup.subtree_control"
+
+
+class CgroupCoreV2(CgroupCore):
+    def is_controller_supported(self, controller: str):
+        return controller in self.read_from_interface_file(CGROUP_SUPPORTED_CONTROLLERS).split()
+
+    def enable_controller(self, controller: str):
+        assert self.is_controller_supported(controller), "Controller not supported"
```

Comment:
```suggestion
        assert self.is_controller_supported(controller), f"Controller not supported {controller!r}"
```
always give the bad data name in exception. this ain't pytest, you won't see the name otherwise